staging/src/k8s.io/apimachinery/pkg/util/runtime/runtime.go

```diff
@@ -161,3 +190,213 @@ func RecoverFromPanic(err *error) {
+	// TODO unexport this field once we change the default value from 1
+	Similar float64
+
+	// clock allows us to control time in unit tests
+	clock clock.Clock
+
+	// logErrorHandler allows us to check err and count in unit tests
+	logErrorHandler func(err error, count uint64)
+
+	// getStackHandler allows us to control the perceived stack in unit tests
+	getStackHandler func() (stack string)
+}
+
+// errKey tracks uniqueness based on the caller's stack and the type/message of the error
+// it is stored in both the count map and the LRU cache
+// it is removed from the count map when it gets evicted from the LRU cache
+// it is comparable via ==
+type errKey struct {
+	stack   string
+	errType reflect.Type
+	// message is the err.Error() value
+	message string
+}
+
+// errVal tracks how many times we have seen an error, and the last time we logged it
+type errVal struct {
+	count  uint64
+	logged time.Time
+}
+
+// handleErr logs the given error if it is considered new or "not recently logged"
+// currently it logs errors whenever:
+// 1. the associated errKey does not exist in d.count (see d.getVal for specifics on how this is calculated)
+// 2. the associated counter is a power of two
+// 3. the associated logged time's delta from the current time is greater than d.delta
+func (d *dedupingErrorHandler) handleErr(err error) {
+	// the operations below that do not acquire the lock do not mutate d
+
+	// we must determine our stack in this function since getStack counts frames
+	stack := d.getStackHandler()
+	key := errKey{stack: stack, errType: reflect.TypeOf(err), message: err.Error()}
+
+	// operations after this point can mutate d
+	// we must acquire the lock before calling getVal since we cannot
+	// have concurrent reads and writes against d.count
+	d.mutex.Lock()
+	defer d.mutex.Unlock()
+
+	// increment our counter
+	val, isNewErr := d.getVal(&key)
+	val.count++
+
+	if isNewErr {
+		// we did not find the error, so add
+		// the associated entry in the LRU cache
+		d.cache.Add(key, nil)
+	} else {
+		// we found this error, so tell the cache that we saw it
+		// cache.Get should always return nil, true
+		d.cache.Get(key)
+	}
+
+	// determine if we need to log this time
+	if isNewErr || isPowerOfTwo(val.count) || d.clock.Since(val.logged) >= d.delta {
+		val.logged = d.clock.Now()
+		d.logErrorHandler(err, val.count)
+	}
+
+	// update the counter in the map after we determine if we need to log it
+	d.count[key] = val
+}
+
+// getVal returns the errVal associated with key, and if the key represents a new error.
+// If the exact key does not exist in d.count, a similar enough fuzzy match on key.message
+// will be used as a fallback (the stack and error type must always match via ==).
+// If fuzzy matching is used, the given key's message will be updated to the similar message.
+// Thus this method requires that key be passed in as a pointer.
+func (d *dedupingErrorHandler) getVal(key *errKey) (errVal, bool) {
```

Comment:
> I would just remove all of the fuzzy matching from this PR and introduce a second PR where that aspect can be focused on.

I really want the fuzzy matching to avoid repeating messages that differ by namespace (this happens fairly frequently).  

@enj How about structuring `getVal` as a settable function value on the struct `getVal func(d *dedupingErrorHandler, key *errKey) (errVal, bool)`.  Create one implementation to use by default here (first few lines of this function).  Downstream, we can cleanly change that default.

@lavalamp fair compromise to get us started?  It makes the review on later changes clean too.

Reply:
That's much more palatable, thanks.